backend/src/room/room.gateway.ts

```diff
@@ -0,0 +1,68 @@
+import { UseGuards } from "@nestjs/common";
+import {
+  ConnectedSocket,
+  MessageBody,
+  SubscribeMessage,
+  WebSocketGateway,
+  WebSocketServer,
+} from "@nestjs/websockets";
+import { InjectPinoLogger, PinoLogger } from "nestjs-pino";
+import { Namespace, Socket } from "socket.io";
+
+import { WsAuthGuard } from "src/auth/ws.guard";
+
+import { RoomService } from "./room.service";
+
+import { ROOM_EVENTS, ROOM_NAMESPACE } from "~shared/constants";
+
+@UseGuards(WsAuthGuard)
+@WebSocketGateway({ namespace: ROOM_NAMESPACE })
+export class RoomGateway {
+  @WebSocketServer()
+  server: Namespace;
+
+  constructor(
+    @InjectPinoLogger()
+    private readonly logger: PinoLogger,
+    private readonly roomService: RoomService,
+  ) {}
+
+  @SubscribeMessage(ROOM_EVENTS.JOIN)
+  async handleJoin(
+    @ConnectedSocket() client: Socket,
+    @MessageBody() roomId: string,
+  ) {
+    this.logger.info(`Joining room: ${client.id}`);
+    const userId = Number(client.handshake.headers.authorization);
+
+    if (await this.roomService.getRoom(userId)) {
+      this.server.to(roomId).emit(ROOM_EVENTS.RECONNECTED);
+    }
+
+    client.join(roomId);
+  }
+
+  @SubscribeMessage(ROOM_EVENTS.LEAVE)
+  async handleLeave(
+    @ConnectedSocket() client: Socket,
+    @MessageBody() roomId: string,
+  ) {
+    const userId = Number(client.handshake.headers.authorization);
+    this.logger.info(`${userId} left rooom`);
+    this.server.to(roomId).emit(ROOM_EVENTS.END_MATCH);
```

Comment:
We should only end the match if both users have left the room, otherwise one user leaving the room might make the other loose his progress unwillingly.

Reply:
Makes sense. The room will only be removed after all users have left